src/Raven.Server/Documents/ShardedHandlers/ShardedHiLoHandler.cs

```diff
@@ -0,0 +1,88 @@
+// -----------------------------------------------------------------------
+//  <copyright file="ShardedHiLoHandler.cs" company="Hibernating Rhinos LTD">
+//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
+//  </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Raven.Client;
+using Raven.Client.Documents.Identity;
+using Raven.Server.Documents.Handlers;
+using Raven.Server.Documents.ShardedHandlers.ShardedCommands;
+using Raven.Server.Documents.Sharding;
+using Raven.Server.Routing;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Documents.ShardedHandlers
+{
+    public class ShardedHiLoHandler : ShardedRequestHandler
+    {
+        [RavenShardedAction("/databases/*/hilo/next", "GET")]
+        public async Task GetNextHiLo()
+        {
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var tag = GetQueryStringValueAndAssertIfSingleAndNotEmpty("tag");
+                var numOfShards = ShardedContext.RequestExecutors.Length;
+                var shardIndex = ShardedContext.GetShardIndex(context, DateTime.UtcNow.Ticks.ToString());
+
+                for (int i = 0; i < numOfShards; i++)
+                {
+                    var cmd = new ShardedCommand(this, Headers.IfMatch);
+
+                    try
+                    {
+                        await ShardedContext.RequestExecutors[shardIndex].ExecuteAsync(cmd, context);
+                    }
+                    catch (Exception)
+                    {
+                        // failed to reach shard
+                        // try the next one
+                        shardIndex = (shardIndex + 1) % numOfShards;
+                        continue;
+                    }
+
+                    HttpContext.Response.StatusCode = (int)cmd.StatusCode;
+
+                    HttpContext.Response.Headers[Constants.Headers.Etag] = cmd.Response?.Headers?.ETag?.Tag;
+
+                    cmd.Result.Modifications = new DynamicJsonValue(cmd.Result)
+                    {
+                        [nameof(HiLoResult.ShardIndex)] = shardIndex
+                    };
+
+                    using (cmd.Result)
+                    using (var result = context.ReadObject(cmd.Result, HiLoHandler.RavenHiloIdPrefix + tag, BlittableJsonDocumentBuilder.UsageMode.ToDisk))
+                    {
+                        await result.WriteJsonToAsync(ResponseBodyStream());
+                    }
+
+                    return;
+                }
+            }
+        }
+
+        [RavenShardedAction("/databases/*/hilo/return", "PUT")]
+        public async Task HiLoReturn()
+        {
+            var shardIndex = GetLongQueryString("shardIndex", required: true);
```

Comment:
I think a better option would be to use sequential shard ids (0..1M) and use their mapping.
In other words, we'll align that to a shard (and ping between nodes as needed), based to the shard topology.
So the shard index won't be the position here, which may change.

Reply:
Another solution might be to move the HiLo doc to be at the system storage of the node.
So it will be node specific only and not shard-specific. All we care here is about non-conflicted ids.
If we do go this way, we need to handle backup/restore properly.

Reply:
That would conflict between different sharded databases, and it doesn't match the way we do everything else.
Better to just split it to a shard like we are always doing.

Reply:
It doesn't match the way we do everything else indeed :-)
but it will not conflict, since it will be done under write lock of the _server_.

Split to shards might be tricky, since they can move from node and from database.

Reply:
That is the _point_, you select a shard id, and return to the shard id.
And the rest is just dispatching to the right node, like we do everywhere here.